### PR TITLE
532 support drug interactions 2

### LINKED ISSUE #532: Support drug interactions
These are supposed to be supported by UC - there is an endpoint for them, but we didn't have a reliable source for the interaction data so it hasn't been completed.

We should complete this, manually storing our own data set of interactions. 

Requirements:
- Add interaction groups
- Add interactions to groups, or individual drugs
- Interactions API request working
- Browsable list of drug interactions (With Login Only)

### PR DESCRIPTION
**Pull request by jmbrunskill**
Fixes #532 

## Description
Adds the ability to Add and Edit Drug interactions

image

image

## Checklist:
- [ ] This PR doesn't make the drug interactions accessible via the V1 API

## Todos:
- [x] Upsert tests (seems like there might be some edge cases missed with switching drugs around)
- [ ] More validations, e.g can't add the same group or drug twice?

### CODE REVIEW

File: backend/dgraph/src/drug_interaction/insert_interaction.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            drugs: vec![DrugCode {
+                code: "294d8414".to_string(),
+            }],
+            groups: vec![
+                InteractionGroupRef {
+                    interaction_group_id: interaction_group1_id.clone(),
+                },
+                InteractionGroupRef {
+                    interaction_group_id: interaction_group2_id.clone(),
+                },
```
Comment (lache-melvin) [inline backend/dgraph/src/drug_interaction/insert_interaction.rs:140]: I mean I guess seeing they're stored as arrays this is possible at the dgraph level... but having the 1 drug and 2 groups is a bit confusing - only 2 things should be interacting right?

File: backend/dgraph/src/drug_interaction/interactions.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+use crate::DrugInteractionsData;
+
+#[derive(Serialize, Debug, Clone)]
+pub struct DrugInteractionFilter {
+    pub name: Option<String>,
+}
+
+pub async fn interactions(
+    client: &DgraphClient,
+    filter: Option<DrugInteractionFilter>,
```
Comment (lache-melvin) [inline backend/dgraph/src/drug_interaction/interactions.rs:14]: NIT but seeing as we've named this `variables` everywhere else, I saw filter and expected that it would be the entire filter struct to pass into the query i.e. `{name: {eq: $name}}`

File: backend/graphql/drug_interactions/src/mutations/delete_interaction.rs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+use service::{
+    auth::{Resource, ResourceAccessRequest},
+    drug_interactions::ModifyDrugInteractionError,
+};
+
+pub async fn delete_drug_interaction(ctx: &Context<'_>, code: String) -> Result<u32> {
+    let user = validate_auth(
+        ctx,
+        &ResourceAccessRequest {
+            resource: Resource::MutateUniversalCodes,
```
Comment (lache-melvin) [inline backend/graphql/drug_interactions/src/mutations/delete_interaction.rs:17]:
```suggestion
            resource: Resource:: MutateDrugInteractions,
```

i think?

File: backend/service/src/drug_interactions/tests/update_interaction.rs [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+        let _result = service
+            .delete_drug_interaction_group(
+                service_provider.clone(),
+                context.user_id.clone(),
+                new_interaction_group_id_2.clone(),
+            )
+            .await
+            .unwrap();
+    }
+}
```
Comment (lache-melvin) [inline backend/service/src/drug_interactions/tests/update_interaction.rs:257]: Might be good to add tests for providing i.e. 2 drugs and a group and also for i.e. 1 drug and no groups?

**Review by lache-melvin — CHANGES_REQUESTED** [review]
Looking awesome! Made a few minor comments but I think main thing to fix is not allowing the same drug/group to be selected twice?

I can select e.g. Abacavir interacts with Abacavir in the modal and it appears to save fine, but upon querying, only drug1 is populated... Do you have time/space to do this or would you like me to?

Will fix up the padding etc. now....

File: frontend/system/src/Admin/Interactions/api/hooks/useUpsertInteraction.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+export const useUpsertDrugInteraction = () => {
+  const { client } = useGql();
+  const sdk = getSdk(client);
+  const queryClient = useQueryClient();
+
+  const invalidateQueries = () => {
+    queryClient.invalidateQueries([DRUG_INTERACTION_KEY]);
+  };
+
+  return [sdk.UpsertDrugInteraction, invalidateQueries] as const;
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply frontend/system/src/Admin/Interactions/api/hooks/useUpsertInteraction.ts:14]: Yeah, I nearly did it today but thought it was faster to copy and paste for now.

File: backend/service/src/drug_interactions/upsert_interaction.rs [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+            .collect(),
+    ))
+}
+
+pub async fn validate(
+    client: &dgraph::DgraphClient,
+    new_item: &UpsertDrugInteraction,
+) -> Result<Option<DrugInteraction>, ModifyDrugInteractionError> {
+    if new_item.name.is_empty() {
+        return Err(ModifyDrugInteractionError::InternalError(
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply backend/service/src/drug_interactions/upsert_interaction.rs:167]: Haha nice, Good point!

File: backend/dgraph/src/drug_interaction/interactions.rs (hunk 2 above)
Comment (jmbrunskill) [reply backend/dgraph/src/drug_interaction/interactions.rs:14]: I had thought I might make it a more fully featured filter, but seeing as I haven't I agree...

Comment (jmbrunskill) [reply backend/dgraph/src/drug_interaction/interactions.rs]: [99309fd](https://github.com/msupply-foundation/unified-codes/pull/574/commits/99309fd5e739ff8298d89547ee07b00936ac4172)

File: backend/service/src/auth.rs [hunk 7]
```diff
@@ -26,6 +26,7 @@ pub enum Resource {
     QueryPendingChanges,
     QueryUsers,
     MutateUniversalCodes,
+    MutateDrugInteractions,
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply backend/service/src/auth.rs:29]: Yeah I was thinking maybe we would but it wasn't really thought out. Was considering the same for barcodes.
Feels like drug interactions might arguably be a bit more high risk than adding a new drug.

File: backend/dgraph/src/drug_interaction/insert_interaction.rs (hunk 1 above)
Comment (jmbrunskill) [reply backend/dgraph/src/drug_interaction/insert_interaction.rs:140]: Agreed, I probably just did this to see if it would work on not then forgot!

Comment (jmbrunskill) [reply backend/dgraph/src/drug_interaction/insert_interaction.rs]: [c022946](https://github.com/msupply-foundation/unified-codes/pull/574/commits/c0229460e4457ae1223ce0b01aac561145c4391c)

File: backend/service/src/drug_interactions/tests/update_interaction.rs (hunk 4 above)
Comment (jmbrunskill) [reply backend/service/src/drug_interactions/tests/update_interaction.rs:392]: [cd81661](https://github.com/msupply-foundation/unified-codes/pull/574/commits/cd816611590f8e2553a7be79aeabae6c66caede7)
